From 6175c9aee103e40b5a5da587f659563de93b3d85 Mon Sep 17 00:00:00 2001
From: Alvin Chang <alvinga@andestech.com>
Date: Thu, 18 Apr 2024 14:52:36 +0800
Subject: [PATCH 226/241] target/riscv: Fix GDB can not read the read-only CSR

From commit 563581cb60, use riscv_csrrw() to read a read-only CSR will
lead to exception. Fix it by calling riscv_csrr() when GDB wants to read
a read-only CSR.

Signed-off-by: Alvin Chang <alvinga@andestech.com>
---
 target/riscv/csr.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/target/riscv/csr.c b/target/riscv/csr.c
index 7aab267916..96accc1549 100644
--- a/target/riscv/csr.c
+++ b/target/riscv/csr.c
@@ -4625,7 +4625,11 @@ RISCVException riscv_csrrw_debug(CPURISCVState *env, int csrno,
 #if !defined(CONFIG_USER_ONLY)
     env->debugger = true;
 #endif
-    ret = riscv_csrrw(env, csrno, ret_value, new_value, write_mask);
+    if (!write_mask) {
+        ret = riscv_csrr(env, csrno, ret_value);
+    } else {
+        ret = riscv_csrrw(env, csrno, ret_value, new_value, write_mask);
+    }
 #if !defined(CONFIG_USER_ONLY)
     env->debugger = false;
 #endif
-- 
2.34.1

